NetKet/Machine/mps_diagonal.hpp

```diff
@@ -0,0 +1,471 @@
+    int k = 0;
+
+    for (int site = 0; site < symperiod_; site++) {
+      for (int spin = 0; spin < d_; spin++) {
+        for (int i = 0; i < D_; i++) {
+          W_[site][spin](i) = pars(k);
+          k++;
+        }
+      }
+    }
+  }
+
+  // Auxiliary function used for setting initial random parameters and adding
+  // identities in every matrix
+  inline void SetParametersIdentity(const VectorType &pars) {
+    int k = 0;
+
+    for (int site = 0; site < symperiod_; site++) {
+      for (int spin = 0; spin < d_; spin++) {
+        for (int i = 0; i < D_; i++) {
+          W_[site][spin](i) = T(1, 0) + pars(k);
+          k++;
+        }
+      }
+    }
+  }
+
+  void InitRandomPars(int seed, double sigma) override {
+    VectorType pars(npar_);
+
+    netket::RandomGaussian(pars, seed, sigma);
+    SetParametersIdentity(pars);
+  }
+
+  int Nvisible() const override { return N_; }
+
+  void InitLookup(const Eigen::VectorXd &v, LookupType &lt) override {
+    // First (left) site
+    _InitLookup_check(lt, 0);
+    lt.V(0) = W_[0][confindex_[v(0)]];
+
+    // Last (right) site
+    _InitLookup_check(lt, 1);
+    lt.V(1) = W_[(N_ - 1) % symperiod_][confindex_[v(N_ - 1)]];
+
+    // Rest sites
+    for (int i = 2; i < 2 * N_; i += 2) {
+      int site = i / 2;
+      _InitLookup_check(lt, i);
+      lt.V(i) =
+          lt.V(i - 2).cwiseProduct(W_[site % symperiod_][confindex_[v(site)]]);
+
+      site = N_ - 1 - site;
+      _InitLookup_check(lt, i + 1);
+      lt.V(i + 1) =
+          W_[site % symperiod_][confindex_[v(site)]].cwiseProduct(lt.V(i - 1));
+    }
+  }
+
+  // Auxiliary function
+  inline void _InitLookup_check(LookupType &lt, int i) {
+    if (lt.VectorSize() == i) {
+      lt.AddVector(D_);
+    } else {
+      lt.V(i).resize(D_);
+    }
+  }
+
+  // Auxiliary function for sorting indeces
+  // (copied from stackexchange - original answer by Lukasz Wiklendt)
+  inline std::vector<std::size_t> sort_indeces(const std::vector<int> &v) {
+    // initialize original index locations
+    std::vector<std::size_t> idx(v.size());
+    std::iota(idx.begin(), idx.end(), 0);
+    // sort indexes based on comparing values in v
+    std::sort(idx.begin(), idx.end(),
+              [&v](std::size_t i1, std::size_t i2) { return v[i1] < v[i2]; });
+    return idx;
+  }
+
+  void UpdateLookup(const Eigen::VectorXd &v, const std::vector<int> &tochange,
+                    const std::vector<double> &newconf,
+                    LookupType &lt) override {
+    std::size_t nchange = tochange.size();
+    if (nchange <= 0) {
+      return;
+    }
+    std::vector<std::size_t> sorted_ind = sort_indeces(tochange);
+    int site = tochange[sorted_ind[0]];
+
+    // Update left (site++)
+    if (site == 0) {
+      lt.V(0) = W_[0][confindex_[newconf[sorted_ind[0]]]];
+    } else {
+      lt.V(2 * site) =
+          lt.V(2 * (site - 1))
+              .cwiseProduct(
+                  W_[site % symperiod_][confindex_[newconf[sorted_ind[0]]]]);
```

Comment:
I think that an elegant and compact way to avoid code duplication is to do something like 
```c++


template<bool B> void MulCwiseOrFull(MatrixType & m1,MatrixType& m2){
  m1*=m2; 
}
template<> void MulCwiseOrFull<true>(MatrixType & m1,MatrixType& m2){
  m1.cwiseProduct(m2); 
}

template<bool IsDiagMps> class Mps{

..... 

void UpdateLookup(...){
  Multiply<IsDiagMps>(M1,M2);
}


}; 
````